csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -301,27 +133,255 @@ std::unordered_set<ParallelType> getParallelTypesToPropagate(
+  std::copy_if(
+      ref->getLoopDomain().begin(),
+      ref->getLoopDomain().end(),
+      std::inserter(device_ids, device_ids.end()),
+      [&selected_parallel_types](IterDomain* id) {
+        return id->isDeviceDim() &&
+            selected_parallel_types.count(id->getParallelType());
+      });
+
+  if (device_ids.empty()) {
+    return;
+  }
+
+  std::unordered_set<IterDomain*> ignored_device_ids;
+  std::unordered_map<IterDomain*, IterDomain*> ref2target;
+  std::unordered_set<IterDomain*> input_ids;
+
+  // Finds all device IDs that are dependent on the given id
+  // and adds them to the ignored_device_ids set.
+  auto add_ignored_device_ids = [&ignored_device_ids,
+                                 &device_ids](IterDomain* id) {
+    for (auto device_id : device_ids) {
+      if (DependencyCheck::isDependencyOf(id, device_id)) {
+        ignored_device_ids.insert(device_id);
+      }
+    }
+  };
+
+  auto is_in_domain = [](IterDomain* id,
+                         const std::vector<IterDomain*>& domain) -> bool {
+    return std::find(domain.begin(), domain.end(), id) != domain.end();
+  };
+
+  if (direction == PropagateDirection::kForward) {
+    input_ids = getInputsInTargetDomain(
+        {device_ids.begin(), device_ids.end()},
+        {ref->getLogicalDomain().begin(), ref->getLogicalDomain().end()});
+    ref2target =
+        PairwiseLogicalDomainMap(ref, tv).mapProducerToConsumer(&input_ids);
+  }
+
+  if (direction == PropagateDirection::kBackward) {
+    input_ids = getInputsInTargetDomain(
+        {device_ids.begin(), device_ids.end()},
+        {ref->getMaybeRootDomain().begin(), ref->getMaybeRootDomain().end()});
+    ref2target =
+        PairwiseLogicalDomainMap(tv, ref).mapConsumerToProducer(&input_ids);
+  }
+
+  for (IterDomain* ref_id : ir_utils::filterByType<IterDomain>(input_ids)) {
+    if (ref2target.find(ref_id) == ref2target.end()) {
+      add_ignored_device_ids(ref_id);
+      continue;
+    }
+
+    IterDomain* target_id = ref2target.at(ref_id);
+
+    // For forward: target_id may not be logical.
+    // For backward: ref_id may not be logical.
+    IterDomain* outermost_ref_id = getOutermostLogicalId(ref_id, ref);
+    IterDomain* outermost_target_id = getOutermostLogicalId(target_id, tv);
+
+    if (outermost_target_id->isParallelized() ||
+        !is_in_domain(outermost_target_id, tv->getLoopDomain())) {
+      // Target ID is further transformed or is already parallelized.
+      add_ignored_device_ids(ref_id);
+      ref2target.erase(ref_id);
+      continue;
+    }
+
+    // Update the ref2target map mapping to/from the outermost logical ID.
+    // instead of the root ID.
+    ref2target.erase(ref_id);
+    ref2target[outermost_ref_id] = outermost_target_id;
+  }
+
+  // Remove device IDs that are not mapped to any target ID,
+  // or map to target IDs that are further transformed.
+  for (IterDomain* device_id : ignored_device_ids) {
+    device_ids.erase(device_id);
```

Comment:
I struggled to understand the computation of device_ids because it's done in several passes. I think we can simplify it as the following:

```
device_ids = []
for ref_loop in ref.loop:
  if doesn't have a selected parallel type:
    continue

  ref_logical = ref_loop walks up  # For simplicity, I don't mind at all if you assume there's only one logical above it.
  target_logical = look up ref_logical in ref2target
  if target_logical is None:
    continue

  if target_logical is already transformed:
    target_logical = find outermost

  if target_logical is not in loop:
    continue

  device_ids.append(...)  
```

I'm sure I'm missing some important details. But the above shows the rough idea.